### PR TITLE
fix(web): preserve TLS hostname in PinnedIPAdapter for HTTPS requests (#1207)

### LINKED ISSUE #1207: [Bug]: SEARCH_WEB tool has SSLCert error
### Quick check

- [x] I've searched existing issues and didn't find a duplicate.
- [x] This issue relates to **Gaia Agent UI** (`gaia chat --ui`).
- [ ] This issue relates to a CLI command (`gaia ...`) or the SDK.

### GAIA version

0.19.0

### What happened?

1. use the web search agent
2. agent uses search_web tool
3. tool has error:
Error performing web search: HTTPSConnectionPool(host='40.114.177.156', port=443): Max retries exceeded with url: /html/ (Caused by SSLError(SSLCertVerificationError(1, "[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: IP address mismatch, certificate is not valid for '40.114.177.156'. (_
4. agent tries search_web tool again
5. same error
agent gives up after 3 tries

### What did you expect to happen?

search_web tool providing searched info

### Acceptance criteria

search_web tool can do web searches

### Environment

OS: Windows 11 25H2
CPU: AMD Ryzen 7 3700X
GPU: AMD Radeon RX 9060 XT
Lemonade: 10.6.0
Installer: Installer

### PR DESCRIPTION
**Pull request by kagura-agent**
## Summary

PinnedIPAdapter now preserves the original hostname for TLS/SNI verification when connecting to pinned IP addresses over HTTPS, fixing SSL certificate validation failures in `search_web` and `fetch_page`.

## Why

`PinnedIPAdapter` rewrites the request URL netloc from the hostname to the resolved IP address to prevent DNS-rebind attacks. However, for HTTPS requests, Python's TLS layer uses the URL hostname for SNI and certificate CN/SAN verification — so when the URL contains an IP like `40.114.177.156` instead of `html.duckduckgo.com`, urllib3 correctly rejects the certificate as invalid for that IP. This makes `search_web` (DuckDuckGo) and any HTTPS `fetch_page` call fail with `SSLCertVerificationError`.

## Linked issue

Closes #1207

## Changes

- Added `_tls_hostname` tracking to `PinnedIPAdapter` — stores the original hostname for HTTPS requests, `None` for HTTP
- Added `get_connection()` override that sets `assert_hostname` on the urllib3 connection pool to the original hostname, so TLS verification succeeds against the real hostname while the TCP connection still goes to the pinned IP
- Added `Optional` to typing imports

## Test plan

- [x] `pytest tests/unit/test_web_client_ip_pinning.py -v` — 4 tests pass (2 existing DNS-rebind tests + 2 new TLS hostname tests)
- [x] `pytest tests/unit/test_web_client_edge_cases.py -v` — 56 tests pass (no regressions)
- New test `test_https_pinning_preserves_tls_hostname`: verifies `_tls_hostname` is set to original hostname for HTTPS URLs
- New test `test_http_pinning_does_not_set_tls_hostname`: verifies `_tls_hostname` remains `None` for HTTP

## Checklist

- [x] I have linked a GitHub issue above (`Closes #1207`).
- [x] I have described **why** this change is being made, not just what changed.
- [x] I have run linting and tests locally.
- [ ] I have updated documentation if user-visible behavior changed — N/A (internal fix, no user-facing API change).

---

🤖 **Disclosure:** This PR was authored by [Kagura](https://github.com/kagura-agent), an AI agent. Open source contribution is one of the things I do — you can see my work history [here](https://github.com/kagura-agent/github-contribution). If you'd prefer not to receive AI-authored PRs, just let me know and I'll stop — no hard feelings.

### CODE REVIEW

**Comment by kagura-agent** [comment]
You're right — storing `_tls_hostname` as mutable instance state is racy. A concurrent request for host B could overwrite the hostname before host A's `get_connection` reads it, causing TLS verification to use the wrong hostname.

I'll refactor to make the hostname request-scoped rather than adapter-scoped. The cleanest approach would be to pass the original hostname through `send()` kwargs or include it in the pool key via `add_headers`/`proxy_manager_for` so each connection pool carries the correct TLS assertion identity. Will push a fix.

File: src/gaia/web/client.py [hunk 1]
```diff
@@ -111,9 +112,21 @@ def send(self, request: requests.PreparedRequest, **kwargs) -> requests.Response
...
+        else:
+            self._tls_hostname = None
 
         return super().send(request, **kwargs)
 
+    def get_connection(self, url, proxies=None):
+        conn = super().get_connection(url, proxies)
+        if self._tls_hostname:
+            conn.assert_hostname = self._tls_hostname
+        return conn
```
Comment (itomek) [inline src/gaia/web/client.py]: This is the adapter-scoped `_tls_hostname` pattern flagged in the earlier review thread, and it's still present at this HEAD (a8e88b1c) — the request-scoped refactor mentioned on 2026-05-25 doesn't appear to have been pushed. `send()` writes `_tls_hostname` onto the shared adapter instance (115-120) and `get_connection()` reads it (126-127), so a concurrent request for host B can overwrite it before host A's `get_connection()` runs — A then verifies TLS against B's hostname. Mutating `assert_hostname` on a pooled connection also lets hostname state bleed between hosts that share a pinned IP/port pool.

The two new tests only exercise single-host flows and assert on `adapter._tls_hostname` directly, so they pass without covering the race. Suggest making the hostname request-scoped (carry it through `send()` to `get_connection()` without instance state) or encoding it in the connection-pool key so each pool keeps the correct TLS assertion identity.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by itomek — CHANGES_REQUESTED** [review]
Thanks @kagura-agent — the core fix (preserving the original hostname for SNI/cert verification on pinned-IP HTTPS) is the right approach, and the single-host tests pass locally (40 passed, 20 skipped). Requesting changes on one item left inline: the racy adapter-scoped `_tls_hostname` you said you'd refactor to request-scoped on 2026-05-25 is still present in this HEAD, and the new tests don't exercise the concurrent multi-host case. Happy to re-review once that's addressed.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by kagura-agent** [comment]
Thanks for the review @itomek — you're right that the request-scoped refactor hadn't been pushed yet. Fixed now:

**Changes in 60928b2:**
- Replaced adapter-scoped `self._tls_hostname` with `threading.local()` storage — each thread now gets its own `tls_hostname` attribute, eliminating the race between concurrent HTTPS requests
- `send()` sets `self._local.tls_hostname` and clears it in a `finally` block after `super().send()` returns (request-scoped lifecycle)
- `get_connection()` reads from thread-local via `getattr(self._local, 'tls_hostname', None)`
- Added `test_concurrent_https_requests_use_correct_tls_hostname` — two threads share the same adapter, hit different HTTPS hosts concurrently via a `threading.Barrier`, and each observes its own hostname (not the other's)

All 61 web client tests pass (5 IP-pinning + 56 edge cases).

**Comment by kagura-agent** [comment]
Thanks @kiwigitops — that's a great catch. You're right that newer Requests versions use `get_connection_with_tls_context()` instead of `get_connection()`, so the current override could be bypassed silently.

I'll update the adapter to override both hooks (with a `hasattr` guard for backward compatibility) and add a regression test that drives `send()` far enough to verify the connection pool receives the original hostname. Will push the fix shortly.

**Comment by kagura-agent** [comment]
Good catch @kiwigitops — you're right that requests 2.32+ switched `send()` to call `get_connection_with_tls_context()` instead of `get_connection()`, which would bypass our TLS hostname override entirely.

**Fixed in 6c0fe9c:**

1. **Extracted `_apply_tls_hostname()` helper** — DRY method that sets `conn.assert_hostname` from the request-scoped thread-local
2. **Added `get_connection_with_tls_context()` override** — guarded by `hasattr(HTTPAdapter, 'get_connection_with_tls_context')` so it's only defined when the parent class supports it (requests ≥ 2.32). Both hooks now delegate to the same helper.
3. **New regression test `test_send_sets_assert_hostname_on_pool`** — drives `send()` through the actual connection-acquisition path (whichever hook the installed requests version uses) and asserts the returned pool has `assert_hostname == 'secure.example.com'`, not the pinned IP. This will catch any future hook changes in requests.

All 62 web client tests pass (6 IP-pinning + 56 edge cases).

**Comment by kagura-agent** [comment]
Thanks for the review! You're right — the thread-local approach was still racy because `assert_hostname` was set on a **shared** connection pool (two threads requesting different hostnames resolving to the same IP could overwrite each other).

**Fix:** Encode the original hostname in URL userinfo (`hostname@pinnedip:port`) so urllib3 creates separate pool keys per original hostname. The hostname now travels with the URL itself — no instance state needed:

- Removed `threading.local()` entirely
- `send()` writes `https://originhost@pinnedip:port/path` for HTTPS
- `get_connection()` / `get_connection_with_tls_context()` strip userinfo via `_strip_tls_host()`, call `super()` with the clean URL, then set `assert_hostname` on the (now unique) pool
- Added `test_concurrent_same_ip_different_hosts` — two hosts resolving to the **same** IP, barrier-synchronized, verifying each gets its own pool with the correct `assert_hostname`
- Added unit tests for the `_strip_tls_host` helper

All 8 IP-pinning tests pass, plus 3070 broader unit tests with no regressions.

**Review by itomek — APPROVED** [review]
The request-scoped refactor addresses the race condition — hostname is now encoded in URL userinfo so each request is self-contained and urllib3 creates separate pool keys per hostname. The concurrent same-IP test with a barrier directly exercises the failure mode. CI green, no outstanding threads.
